torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -451,77 +456,212 @@ void FusionKernelRuntime::prepareRuntimeOrder() {
   }
 }
 
-std::vector<at::Tensor> FusionKernelRuntime::runWithInput(
-    const at::ArrayRef<IValue>& inputs,
-    size_t input_id) {
-  FUSER_PERF_SCOPE("FusionKernelRuntime::runMultiKernelWithInput");
+// passing args by value, since we will be modify this
+void FusionKernelRuntime::compileAsync(KernelArgumentHolder& args_old) {
+  // only single compilation is supported at this moment.
+  std::unique_lock<std::mutex> unique_lock(mutex_, std::try_to_lock);
+  TORCH_CHECK(
+      unique_lock.owns_lock(),
+      "Calling compileAsync on a FusionKernelRuntime that's already starting a compilation thread is not supported");
+  std::unique_lock<std::mutex> unique_lock2(compiling_, std::try_to_lock);
+  TORCH_CHECK(
+      unique_lock2.owns_lock(),
+      "Calling compileAsync on a FusionKernelRuntime that's already starting a compilation thread is not supported 2");
+
+  // for some reason I can't seem to move unique_lock and it keeps using copy.
```

Comment:
I don't *have* to move the unique_lock around. I can directly acquire the mutex at the beginning of the function and release it at the end of the lambda, but I'm trying to avoid calling try_to_lock/unlock directly.

I do need to rethink about the mutex uses here. FusionKernelRuntime can be queried while it's under compilation, so we definitely want a mutex guard somewhere to avoid racing, but probably not during the whole compilation process. Right now I'm just trying to be more conservative and not have a hang in my test.

Reply:
> but I'm trying to avoid calling try_to_lock/unlock directly

why is that?

Reply:
Just trying to follow good practice and use RAII more :)